<font color="green">*To start working on this notebook, or any other notebook that we will use in the Moringa Data Science Course, we will need to save our own copy of it. We can do this by clicking File > Save a Copy in Drive. We will then be able to make edits to our own copy of this notebook.*</font>

# Python Programming: P-values Exercise

## Question 1

The Ministry of Roads and Transportation requires tire manufacturers to provide performance information on tire sidewalls to help prospective buyers make their purchasing decisions. One very important piece of information is the tread wear index, which indicates the tire’s resistance to tread wear. A tire with a grade of 200 should last twice as long, on average, as a tire with a grade of 100.
A consumer organization wants to test the actual tread wear index of a brand name of tires that claims “graded 200” on the sidewall of the tire. A random sample of n = 18 indicates a sample mean tread wear index of 195.3 and a sample standard devia- tion of 21.4.

1.   Is there evidence that the population mean tread wear index is different from 200? (Use a 0.05 level of significance.)
2.   Determine the p-value and interpret its meaning.


In [22]:
#https://www.analyticsvidhya.com/blog/2021/07/t-test-performing-hypothesis-testing-with-python/
# Systolic blood pressures of 14 patients are given below:

# 183, 152, 178, 157, 194, 163, 144, 114, 178, 152, 118, 158, 172, 138

# Test, whether the population mean, is less than 165

# Hypothesis
# H0: There is no significant mean difference in systolic blood pressure. i.e., μ = 165

# H1: The population mean is less than 165. i.e., μ < 165


sys_bp=[183, 152, 178, 157, 194, 163, 144, 114, 178, 152, 118, 158, 172, 138]


mu = 165

from scipy import stats

t_value,p_value=stats.ttest_1samp(sys_bp,mu)


one_tailed_p_value=float("{:.6f}".format(p_value/2)) # Since alternative hypothesis is one tailed, We need to divide the p value by 2.

print('Test statistic is %f'%float("{:.6f}".format(t_value)))

print('p-value for one tailed test is %f'%one_tailed_p_value)

alpha = 0.05

if one_tailed_p_value<=alpha:

    print('Conclusion','n','Since p-value(=%f)'%p_value,'<','alpha(=%.2f)'%alpha,'''We reject the null hypothesis H0. 

So we conclude that there is no significant mean difference in systolic blood pressure. 

i.e., μ = 165 at %.2f level of significance'''%alpha)

else:
    
    print('Conclusion','n','Since p-value(=%f)'%one_tailed_p_value,'>','alpha(=%.2f)'%alpha)



Test statistic is -1.243183
p-value for one tailed test is 0.117877
Conclusion n Since p-value(=0.117877) > alpha(=0.05)


In [41]:
# t Test for Hypothesis of the Mean
#

# 
#OUR CODE GOES HERE
# Find the z score 
n = 18
t_mean = 195.3
t_std = 21.4
zscore = (n-t_mean)/t_std

#Find the p-value 
#import norm function from the scipy library
from scipy import stats
import scipy.stats
from scipy.stats import norm
prob = stats.norm.cdf(zscore)
prob


print("the z-score is " + str(zscore))
print("the P value is " + str(prob))


the z-score is -8.285046728971963
the P value is 5.903136010034462e-17


## Question 2

The following data represents the amount of soft drink filled in a sample of 50 consecutive 2-liter bottles. The results, listed horizontally in the order of being filled, were:

2.109 2.086 2.036 2.038 2.015 2.014 2.005 2.003 1.984 1.981 1.963 1.957 1.908 1.894 2.066 2.075 2.031 2.029 2.013 2.014 1.999 1.996 1.973 1.975 1.951 1.951 2.065 2.057 2.025 2.029 2.012 2.012 1.997 1.992 1.971 1.969 1.947 1.941 2.052 2.044 2.023 2.020 2.012 2.010 1.994 1.986 1.966 1.967 1.941 1.938

At the 0.05 level of significance, is there evidence that the mean amount of soft drink filled is different from 2.0 liters?
Determine the p-value in (1) and interpret its meaning.
In (1), you assumed that the distribution of the amount of soft drink filled was normally distributed. Evaluate this assumption by constructing a boxplot or a normal probability plot.
Do you think that the assumption needed in order to conduct the t test in (1) is valid? Explain.
Examine the values of the 50 bottles in their sequential order, as given in the problem. Does there appear to be a pattern to the results? If so, what impact might this pattern have on the validity of the results in (1)?

In [43]:
drinks = "2.109 2.086 2.036 2.038 2.015 2.014 2.005 2.003 1.984 1.981 1.963 1.957 1.908 1.894 2.066 2.075 2.031 2.029 2.013 2.014 1.999 1.996 1.973 1.975 1.951 1.951 2.065 2.057 2.025 2.029 2.012 2.012 1.997 1.992 1.971 1.969 1.947 1.941 2.052 2.044 2.023 2.020 2.012 2.010 1.994 1.986 1.966 1.967 1.941 1.938"
data = drinks.split()
#We should use z-test ssince number is more than 30
drinks_data = [float(i) for i in data]
print(len(drinks_data))
mu = 2.0
t_value,p_value=stats.ttest_1samp(drinks_data,mu)

print("the P-Value is "+ str(p_value))
print(drinks_data)
#Since the p-value is more than the significant level, we accept the null hypothesis that the mean is 2.0l


50
the P-Value is 0.9095105442498366
[2.109, 2.086, 2.036, 2.038, 2.015, 2.014, 2.005, 2.003, 1.984, 1.981, 1.963, 1.957, 1.908, 1.894, 2.066, 2.075, 2.031, 2.029, 2.013, 2.014, 1.999, 1.996, 1.973, 1.975, 1.951, 1.951, 2.065, 2.057, 2.025, 2.029, 2.012, 2.012, 1.997, 1.992, 1.971, 1.969, 1.947, 1.941, 2.052, 2.044, 2.023, 2.02, 2.012, 2.01, 1.994, 1.986, 1.966, 1.967, 1.941, 1.938]


In [31]:
#From the drinks data take 10 and compare if there is difference 
import numpy as np 
sample_size = 10
bottle_sample = np.random.choice(drinks_data, sample_size)
bottle_sample

array([2.014, 2.052, 1.992, 1.994, 2.01 , 2.01 , 1.941, 1.981, 1.941,
       2.086])

In [32]:
drinks_mean=np.mean(drinks_data)
sample_mean=np.mean(bottle_sample)

In [37]:
#checking if they have significant difference 
t_value,p_value=stats.ttest_1samp(bottle_sample,drinks_mean)
t_value,p_value

(0.09792160668271567, 0.9241406788245341)

In [42]:
# def systematic_sampling(df,r,n):
#     step=df.shape[0]//n
#     indexes = np.arange(start=r, stop= len(df), step=step)
#     systematic_sample = df[indexes]
#     return systematic_sample#, indexes
# systematic_sampling(bottle_df,0,10)


## Question 3

The population mean waiting time to check out of a supermarket has been 4 minutes. recently, in an effort to reduce the waiting time, the supermarket has experimented with a system in which infrared cameras use body heat and in-store software to determine how many lanes should be opened. A sample of 100 customers was selected, and their mean waiting time to check out was 3.25 minutes, with a sample standard deviation of 2.7 minutes.

1. At the 0.05 level of significance, using the p-value approach to hypothesis testing, is there evidence that the population mean waiting time to check out is less than 4 minutes?
2. Interpret the meaning of the p-value in this problem.  

In [ ]:
mu = 3.25



## Question 4

A survey of nonprofit organizations showed that online fun- draising has increased in the past year. Based on a random sample of 55 nonprofit organizations, the mean one-time gift donation in the past year was Ksh. 750, with a standard deviation of Ksh. 900.

1. If you test the null hypothesis at the 0.01 level of significance, is there evidence that the mean one-time gift donation is greater than Ksh. 700?
2. Interpret the meaning of the p-value in this problem.

In [ ]:
OUR CODE GOES HERE

## Question 5

A problem with a phone line that prevents a customer from receiving or making calls is upsetting to both the customer and the telecommunications company. The file Phone contains samples of 20 problems reported to two different offices of a telecommuni- cations company and the time to clear these problems (in minutes) from the customers’ lines:

Central Office I Time to Clear Problems (minutes)
1.48 1.75 0.78 2.85 0.52 1.60 4.15 3.97 1.48 3.10 1.02 0.53 0.93 1.60 0.80 1.05 6.32 3.93 5.45 0.97

Central Office II Time to Clear Problems (minutes)
7.55 3.75 0.10 1.10 0.60 0.52 3.30 2.10 0.58 4.02 3.75 0.65 1.92 0.60 1.53 4.23 0.08 1.48 1.65 0.72

1. Assuming that the population variances from both offices are equal, is there evidence of a difference in the mean waiting time between the two offices? (Use a = 0.05.)
2. Find the p-value in (1) and interpret its meaning.
3. What other assumption is necessary in (1)?
4. Assuming that the population variances from both offices are equal, construct and interpret a 95% confidence interval estimate of the difference between the population means in the two offices.

In [ ]:
OUR CODE GOES HERE

## Question 6

A bank with a branch located in a commercial district of a city has the business objective of developing an improved process for serving customers during the noon-to-1 p.m. lunch period. Management decides to first study the waiting time in the current process. The waiting time is defined as the number of minutes that elapses from when the customer enters the line until he or she reaches the teller window. Data are collected from a random sample of 15 customers and stored in Bank1. 

These data are:
4.21 5.55 3.02 5.13 4.77 2.34 3.54 3.20 4.50 6.10 0.38 5.12 6.46 6.19 3.79

Suppose that another branch, located in a residential area, is also concerned with improving the process of serving customers in the noon-to-1 p.m. lunch period. Data are collected from a random sample of 15 customers and stored in Bank2. 

These data are:
9.66 5.90 8.02 5.79 8.73 3.82 8.01 8.35 10.49 6.68 5.64 4.08 6.17 9.91 5.47


1. Assuming that the population variances from both banks are equal, is there evidence of a difference in the mean waiting time between the two branches? (Use a = 0.05.)
2. Determine the p-value in (a) and interpret its meaning.
3. In addition to equal variances, what other assumption is neces-
sary in (a)?
4. Construct and interpret a 95% confidence interval estimate
of the difference between the population means in the two branches.

In [ ]:
OUR CODE GOES HERE